File: sdk/runanywhere-flutter/packages/runanywhere/lib/public/api/types/model_registration.dart
// SPDX-License-Identifier: Apache-2.0
//
// Registration and filter inputs for `RunAnywhere.models`.

import 'package:runanywhere/generated/model_types.pb.dart'
    show ModelFileDescriptor, ModelQuery;
import 'package:runanywhere/generated/model_types.pbenum.dart'
    show
        ArchiveStructure,
        ArchiveType,
        InferenceFramework,
        ModelArtifactType,
        ModelCategory,
        ModelSource;

/// How a model's files are laid out at its source.
enum ModelArtifactKind {
  /// One downloadable weight file.
  singleFile,

  /// One downloadable archive that unpacks into a model directory.
  archive,

  /// Several files that must all be present.
  multiFile,
}

/// A model to add to the registry: single file, archive, or multi-file.
class ModelRegistration {
  const ModelRegistration._({
    required this.kind,
    required this.name,
    required this.framework,
    required this.category,
    this.id,
    this.url,
    this.archiveStructure,
    this.archiveType,
    this.files = const <ModelFileDescriptor>[],
    this.artifactType,
    this.memoryRequirementBytes,
    this.downloadSizeBytes,
    this.contextLength,
    this.supportsThinking = false,
    this.supportsLora = false,
    this.source = ModelSource.MODEL_SOURCE_REMOTE,
    this.description,
    this.cuaProfile,
  });

  /// Register one downloadable weight file.
  factory ModelRegistration.url({
    required String name,
    required String url,
    required InferenceFramework framework,
    String? id,
    ModelCategory category = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    ModelArtifactType? artifactType,
    int? memoryRequirementBytes,
    int? downloadSizeBytes,
    int? contextLength,
    ModelSource source = ModelSource.MODEL_SOURCE_REMOTE,
    String? description,
    bool supportsThinking = false,
    bool supportsLora = false,
    String? cuaProfile,
  }) => ModelRegistration._(
    kind: ModelArtifactKind.singleFile,
    name: name,
    framework: framework,
    category: category,
    id: id,
    url: url,
    artifactType: artifactType,
    memoryRequirementBytes: memoryRequirementBytes,
    downloadSizeBytes: downloadSizeBytes,
    contextLength: contextLength,
    source: source,
    description: description,
    supportsThinking: supportsThinking,
    supportsLora: supportsLora,
    cuaProfile: cuaProfile,
  );

  /// Register an archive that unpacks into a model directory.
  factory ModelRegistration.archive({
    required String name,
    required String url,
    required ArchiveStructure structure,
    required InferenceFramework framework,
    String? id,
    ModelCategory category = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    ArchiveType? archiveType,
    int? memoryRequirementBytes,
    bool supportsThinking = false,
    bool supportsLora = false,
    String? cuaProfile,
  }) => ModelRegistration._(
    kind: ModelArtifactKind.archive,
    name: name,
    framework: framework,
    category: category,
    id: id,
    url: url,
    archiveStructure: structure,
    archiveType: archiveType,
    memoryRequirementBytes: memoryRequirementBytes,
    supportsThinking: supportsThinking,
    supportsLora: supportsLora,
    cuaProfile: cuaProfile,
  );

  /// Register several files that together make up one model.
  factory ModelRegistration.multiFile({
    required String id,
    required String name,
    required List<ModelFileDescriptor> files,
    required InferenceFramework framework,
    ModelCategory category = ModelCategory.MODEL_CATEGORY_LANGUAGE,
    int? memoryRequirementBytes,
    int? downloadSizeBytes,
    int? contextLength,
    bool supportsThinking = false,
    ModelSource source = ModelSource.MODEL_SOURCE_REMOTE,
    String? cuaProfile,
  }) => ModelRegistration._(
    kind: ModelArtifactKind.multiFile,
    name: name,
    framework: framework,
    category: category,
    id: id,
    files: files,
    memoryRequirementBytes: memoryRequirementBytes,
    downloadSizeBytes: downloadSizeBytes,
    contextLength: contextLength,
    supportsThinking: supportsThinking,
    source: source,
    cuaProfile: cuaProfile,
  );

  /// Which registration shape this is.
  final ModelArtifactKind kind;

  /// Display name.
  final String name;

  /// Engine that will run the model.
  final InferenceFramework framework;

  /// Modality the model serves.
  final ModelCategory category;

  /// Registry id. Derived from [name] when omitted.
  final String? id;

  /// Download URL for single-file and archive registrations.
  final String? url;

  /// On-disk layout the archive unpacks into.
  final ArchiveStructure? archiveStructure;

  /// Archive compression, when it cannot be inferred from [url].
  final ArchiveType? archiveType;

  /// Files for a multi-file registration.
  final List<ModelFileDescriptor> files;

  /// Artifact type hint for single-file registrations.
  final ModelArtifactType? artifactType;

  /// Resident memory the model needs.
  final int? memoryRequirementBytes;

  /// Artifact bytes used for download planning and validation.
  final int? downloadSizeBytes;

  /// Context window the model supports.
  final int? contextLength;

  /// True when the model emits chain-of-thought.
  final bool supportsThinking;

  /// True when LoRA adapters may be applied.
  final bool supportsLora;

  /// Where the model comes from.
  final ModelSource source;

  /// Optional catalog description preserved on the registry entry.
  final String? description;

  /// Computer-Use-Agent profile id (e.g. `'fara'`) for a CUA-capable model.
  /// Lands on `ModelInfo.cuaProfile` so callers can discover which
  /// registered models are drivable through `RunAnywhere.cua`.
  final String? cuaProfile;
}

/// Narrows `models.list` to a subset of the registry.
class ModelFilter {
  /// Filter by any combination of modality, engine, and download state.
  const ModelFilter({
    this.category,
    this.framework,
    this.downloadedOnly,
    this.searchQuery,
  });

  /// Keep only models serving this modality.
  final ModelCategory? category;

  /// Keep only models runnable by this engine.
  final InferenceFramework? framework;

  /// True keeps only models already on disk.
  final bool? downloadedOnly;

  /// Free-text match against id and name.
  final String? searchQuery;

  /// Build the generated registry query.
  ModelQuery toProto() {
    final query = ModelQuery();
    if (category != null) query.category = category!;
    if (framework != null) query.framework = framework!;
    if (downloadedOnly != null) query.downloadedOnly = downloadedOnly!;
    final search = searchQuery;
    if (search != null && search.isNotEmpty) query.searchQuery = search;
    return query;
  }
}

File: bindings/swift/Sources/RunAnywhere/Public/API/ModelRegistration.swift
//
//  ModelRegistration.swift
//  RunAnywhere
//

/// One builder covering URL, archive, and multi-file model registration.
public struct ModelRegistration: Sendable {

    enum Payload: Sendable {
        case url(String)
        case archive(url: String, structure: RAArchiveStructure, type: RAArchiveType?)
        case multiFile([RAModelFileDescriptor])
    }

    let payload: Payload
    public var id: String?
    public var name: String
    public var framework: InferenceFramework
    public var category: ModelCategory
    public var memoryRequirementBytes: Int64?
    public var downloadSizeBytes: Int64?
    public var contextLength: Int?
    public var source: ModelSource
    public var description: String?
    public var supportsThinking: Bool
    public var supportsLora: Bool

    /// Computer-Use-Agent profile id (e.g. `RunAnywhere.CUA.faraProfile`) for a
    /// CUA-capable model; lands on `RAModelInfo.cuaProfile` so callers can
    /// discover which registered models are drivable through `RunAnywhere.CUA`.
    public var cuaProfile: String?

    private init(
        payload: Payload,
        id: String?,
        name: String,
        framework: InferenceFramework,
        category: ModelCategory,
        memoryRequirementBytes: Int64?,
        downloadSizeBytes: Int64?,
        contextLength: Int?,
        source: ModelSource,
        description: String?,
        supportsThinking: Bool,
        supportsLora: Bool,
        cuaProfile: String? = nil
    ) {
        self.payload = payload
        self.id = id
        self.name = name
        self.framework = framework
        self.category = category
        self.memoryRequirementBytes = memoryRequirementBytes
        self.downloadSizeBytes = downloadSizeBytes
        self.contextLength = contextLength
        self.source = source
        self.description = description
        self.supportsThinking = supportsThinking
        self.supportsLora = supportsLora
        self.cuaProfile = cuaProfile
    }

    /// Register a model served from a single download URL or `hf.co` reference.
    public static func url(
        _ url: String,
        name: String,
        framework: InferenceFramework,
        category: ModelCategory = .language,
        id: String? = nil,
        memoryRequirementBytes: Int64? = nil,
        downloadSizeBytes: Int64? = nil,
        contextLength: Int? = nil,
        source: ModelSource = .remote,
        description: String? = nil,
        supportsThinking: Bool = false,
        supportsLora: Bool = false,
        cuaProfile: String? = nil
    ) -> ModelRegistration {
        ModelRegistration(
            payload: .url(url),
            id: id,
            name: name,
            framework: framework,
            category: category,
            memoryRequirementBytes: memoryRequirementBytes,
            downloadSizeBytes: downloadSizeBytes,
            contextLength: contextLength,
            source: source,
            description: description,
            supportsThinking: supportsThinking,
            supportsLora: supportsLora,
            cuaProfile: cuaProfile
        )
    }

    /// Register an archive-packaged model whose on-disk layout the URL cannot imply.
    public static func archive(
        _ url: String,
        structure: RAArchiveStructure,
        name: String,
        framework: InferenceFramework,
        category: ModelCategory = .language,
        archiveType: RAArchiveType? = nil,
        id: String? = nil,
        memoryRequirementBytes: Int64? = nil,
        downloadSizeBytes: Int64? = nil,
        contextLength: Int? = nil,
        source: ModelSource = .remote,
        description: String? = nil,
        supportsThinking: Bool = false,
        supportsLora: Bool = false,
        cuaProfile: String? = nil
    ) -> ModelRegistration {
        ModelRegistration(
            payload: .archive(url: url, structure: structure, type: archiveType),
            id: id,
            name: name,
            framework: framework,
            category: category,
            memoryRequirementBytes: memoryRequirementBytes,
            downloadSizeBytes: downloadSizeBytes,
            contextLength: contextLength,
            source: source,
            description: description,
            supportsThinking: supportsThinking,
            supportsLora: supportsLora,
            cuaProfile: cuaProfile
        )
    }

    /// Register a model made of several files, each carrying its own URL.
    public static func multiFile(
        _ files: [RAModelFileDescriptor],
        id: String,
        name: String,
        framework: InferenceFramework,
        category: ModelCategory = .language,
        memoryRequirementBytes: Int64? = nil,
        downloadSizeBytes: Int64? = nil,
        contextLength: Int? = nil,
        source: ModelSource = .remote,
        description: String? = nil,
        supportsThinking: Bool = false,
        supportsLora: Bool = false,
        cuaProfile: String? = nil
    ) -> ModelRegistration {
        ModelRegistration(
            payload: .multiFile(files),
            id: id,
            name: name,
            framework: framework,
            category: category,
            memoryRequirementBytes: memoryRequirementBytes,
            downloadSizeBytes: downloadSizeBytes,
            contextLength: contextLength,
            source: source,
            description: description,
            supportsThinking: supportsThinking,
            supportsLora: supportsLora,
            cuaProfile: cuaProfile
        )
    }
}
